---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# sizeR <img src="man/figures/README-sizeR_logo.svg" align="right" height="138" /></a>

<!-- badges: start -->
[![Lifecycle: experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://lifecycle.r-lib.org/articles/stages.html#experimental)
<!-- badges: end -->

## Overview

The goal of `sizeR` is to provide a collection of computational tools to analyze gene sizes within gene features (e.g. expression) and gene sets (e.g. upregulated/downregulated).

If you use `sizeR`, please cite our [paper]():
```
sizeR: an R package for the analysis of gene size within gene features and gene sets
MJ McCoy and AZ Fire
XXX, Volume XX, Issue XX, Month 2024, DOI, URL
```

## Requirements

The `sizeR` package works with R version >= 3.6.3 and several packages from `tidyverse`. Either `tidyverse` can be installed and loaded, or these individual packages can be installed and loaded:

* `dplyr` version >= 1.0.9
* `tidyr` version >= 1.2.0
* `readr` version >= 2.1.2
* `ggplot2` version >= 3.3.6
* `Hmisc` version >= 5.1.1

## Installation

You can install the development version of `sizeR` from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("mjmccoy/sizeR")
```

## Example usage

This is series of examples which shows how to perform gene size enrichment analysis with either quantitative or categorical variables. The example data was modified from the following paper:

```
Heat shock induces premature transcript termination and reconfigures the human transcriptome.
S Cugusi, M Richard, K Gavin, J Walker, Z Han, P Pisano, M Wierer, A Stewart, and J Svejstrup
Molecular Cell 82, 1573-1588.e10 (2022), DOI: 10.1016/j.molcel.2022.01.007
```

### Load libraries
```{r libraries, message=FALSE, warning=FALSE}
library(dplyr)
library(tidyr)
library(readr)
library(Hmisc)
library(ggplot2)
library(sizeR)
```

### Load user-specified gene coordinates

Gene coordinates can be extracted from an annotation file (e.g. gtf, gff) or downloaded from [Ensembl](https://www.ensembl.org/biomart/martview). Here we use gene coordinates from Homo sapiens GRCh38.p14 Ensembl release 110. These should be in the format chromosome (e.g. Chromosome/scaffold name), chromosome start position [e.g. Gene start (bp)], chromosome end position [e.g. Gene end (bp)], and gene_id (e.g. Gene stable ID), in this specific order.

```{r gene lengths, message=FALSE, warning=FALSE}
# Load user-provided gene coordinates and specify name of gene_id in the gene coordinates file (in this example: Gene stable ID)
lengths.df <- gene_lengths(
  filepath = "inst/extdata/hsap_GRCh38.p14_ensembl_release_110_gene_lengths.txt",
  gene_id = "Gene stable ID",
  delim = "\t")

head(lengths.df)
```

### Estimate gene size enrichment for quantitative variables
#### Load quantitative data
```{r quantitative variables, message=FALSE, warning=FALSE}
data.df <- sizeR_input(
  file = "inst/extdata/example_expression_data.tsv",
  gene_id = "gene_id",
  delim = "\t")

head(data.df)
```

#### Append gene lengths
```{r}
data.df <- add_lengths(data.df, lengths.df)

head(data.df)
```

#### Perform binomial test for gene size enrichment/depletion
```{r}
binomial.df <- binomial_test(data.df, feature_name = "log2FC")

head(binomial.df)
```

#### sizeR tile plot
```{r}
sizeR_plot(binomial.df, type = "tile", quantiles = FALSE)
```

#### sizeR bar plot
```{r fig.height=2, fig.width=16}
sizeR_plot(binomial.df, type = "bar")
```

### Estimate gene size enrichment for quantitative variables by sample
#### Load quantitative data
```{r quantitative variable by sample, fig.height=4, fig.width=12, message=FALSE, warning=FALSE}
data.df <- sizeR_input(
  file = "inst/extdata/example_expression_by_sample_data.tsv",
  gene_id = "Gene stable ID",
  delim = "\t")

head(data.df)
```

#### Append gene lengths
```{r}
data.df <- add_lengths(data.df, lengths.df)

head(data.df)
```

#### Perform binomial test for gene size enrichment/depletion
```{r}
# Binomial test
binomial.df <- binomial_test(data.df, by_sample = T)

head(binomial.df)
```

#### Plot sizeR tile plot
```{r fig.height=4, fig.width=12}
sizeR_plot(binomial.df, by_sample = T, type = "tile")
```

#### Plot sizeR bar plot
```{r fig.height=8, fig.width=12}
sizeR_plot(binomial.df, by_sample = T, type = "bar")
```

#### Plot sizeR line plot
```{r sizeR_plot line, message=FALSE, warning=FALSE}
sizeR_plot(data.df, type = "line", quantiles = FALSE) +
  scale_color_manual(values = c("condition1.r1" = "firebrick", "condition1.r2" = "firebrick", "condition2.r1" = "black", "condition2.r2" = "black"))
```

### Estimate gene size enrichment for categorical variables
#### Load gene set data
```{r message=FALSE, warning=FALSE}
# Load gene set data
data.df <- sizeR_input(
  file = "inst/extdata/example_gene_set_data.tsv",
  gene_id = "Gene stable ID",
  delim = "\t")

head(data.df)
```

#### Append gene lengths
```{r message=FALSE, warning=FALSE}
data.df <- add_lengths(data.df, lengths.df)

head(data.df)
```

#### Perform binomial test for gene size enrichment/depletion
```{r message=FALSE, warning=FALSE}
binomial.df <- binomial_test(data.df, categorical = T)

head(binomial.df)
```

#### Plot sizeR tile plot
```{r message=FALSE, warning=FALSE}
sizeR_plot(binomial.df, categorical = T, type = "tile")
```

#### Plot sizeR bar plot
```{r fig.height=4, fig.width=8}
sizeR_plot(binomial.df, categorical = T, type = "bar")
```

# Contact
If you have any comments or suggestions please raise an issue or contact us:\
Matthew McCoy: mjmccoy@stanford.edu
